lib/segment/src/id_tracker/immutable_id_tracker.rs

```diff
@@ -0,0 +1,429 @@
+        Box::new(
+            (0..self.mappings.internal_to_external.len() as PointOffsetType)
+                .filter(move |i| !self.deleted[*i as usize]),
+        )
+    }
+
+    fn iter_from(
+        &self,
+        external_id: Option<PointIdType>,
+    ) -> Box<dyn Iterator<Item = (PointIdType, PointOffsetType)> + '_> {
+        let full_num_iter = || {
+            self.mappings
+                .external_to_internal
+                .iter()
+                .filter(|i| i.0.is_num_id())
+                .map(|(k, v)| (k.into(), *v))
+        };
+        let offset_num_iter = |offset: u64| {
+            self.mappings
+                .external_to_internal
+                .range(StoredPointId::NumId(offset)..)
+                .filter(|i| i.0.is_num_id())
+                .map(|(k, v)| (k.into(), *v))
+        };
+        let full_uuid_iter = || {
+            self.mappings
+                .external_to_internal
+                .iter()
+                .filter(|i| i.0.is_uuid())
+                .map(|(k, v)| (k.into(), *v))
+        };
+        let offset_uuid_iter = |offset: Uuid| {
+            self.mappings
+                .external_to_internal
+                .range(StoredPointId::Uuid(offset)..)
+                .filter(|i| i.0.is_uuid())
+                .map(|(k, v)| (k.into(), *v))
+        };
+
+        match external_id {
+            None => {
+                let iter_num = full_num_iter();
+                let iter_uuid = full_uuid_iter();
+                // order is important here, we want to iterate over the u64 ids first
+                Box::new(iter_num.chain(iter_uuid))
+            }
+            Some(offset) => match offset {
+                PointIdType::NumId(idx) => {
+                    // Because u64 keys are less that uuid key, we can just use the full iterator for uuid
+                    let iter_num = offset_num_iter(idx);
+                    let iter_uuid = full_uuid_iter();
+                    // order is important here, we want to iterate over the u64 ids first
+                    Box::new(iter_num.chain(iter_uuid))
+                }
+                PointIdType::Uuid(uuid) => {
+                    // if offset is a uuid, we can only iterate over uuids
+                    Box::new(offset_uuid_iter(uuid))
+                }
+            },
+        }
+    }
+
+    fn iter_ids(&self) -> Box<dyn Iterator<Item = PointOffsetType> + '_> {
+        self.iter_internal()
+    }
+
+    /// Creates a flusher function, that flushes the deleted points bitvec to disk.
+    fn mapping_flusher(&self) -> Flusher {
+        if !*self.dirty.read() {
+            return Box::new(|| Ok(()));
+        }
+
+        let deleted = self.deleted.clone();
+        let path = self.deleted_file_path();
+        let dirty = self.dirty.clone();
+        Box::new(move || {
+            Self::save_file(&path, &deleted)?;
+            *dirty.write() = false;
+            Ok(())
+        })
+    }
```

Comment:
If we move to the approach using memory mapping, we can remove the dirty flag and simply return it's associated flusher.